code/modules/research/xenobiology/crossbreeding/warping.dm

```diff
@@ -0,0 +1,1227 @@
+	move_resist = INFINITY
+	anchored = TRUE
+	///where the rune will teleport you back.
+	var/turf/exit_turf
+	///rune linked to the exit rune
+	var/obj/effect/warped_rune/rainbowspace/enter_rune
+
+
+/datum/map_template/warped_room
+	name = "Warped room"
+	mappath = '_maps/templates/warped_room.dmm'
+
+
+/area/warped_room
+	name = "warped room"
+	icon_state = "yellow"
+	dynamic_lighting = DYNAMIC_LIGHTING_FORCED
+	requires_power = FALSE
+	has_gravity = TRUE
+	noteleport = TRUE
+
+
+///creates the warped room and place an exit rune to exit the room
+/obj/effect/warped_rune/rainbowspace/Initialize()
+	. = ..()
+	rune_room = new()
+	room_reservation = SSmapping.RequestBlockReservation(rune_room.width, rune_room.height) //monkey sees valid location
+	rune_room.load(locate(room_reservation.bottom_left_coords[1], room_reservation.bottom_left_coords[2], room_reservation.bottom_left_coords[3]))//monkey room activate
+	var/obj/effect/warped_room_exit/exit_rune = new(locate(room_reservation.bottom_left_coords[1] + 3, room_reservation.bottom_left_coords[2] + 6, room_reservation.bottom_left_coords[3]))
+	exit_rune.exit_turf = rune_turf //we spawn the exit rune manually in the code because it's easier to link it to the rainbow rune that way.
+	exit_rune.enter_rune = src
+
+
+///here to check if anyone's being transported in or out of the room with the user.
+/obj/effect/warped_rune/rainbowspace/proc/customer_check(atom/person_checked ,smuggle_in)
+	var/list/hidden_customers = person_checked.GetAllContents(/mob/living/carbon/human)
+	if(!length(hidden_customers))
+		return
+	for(var/mob/living/carbon/human/customer in hidden_customers)
+		if(smuggle_in)
+			LAZYADD(customer_list, customer) //if they enter the room
+		else
+			LAZYREMOVE(customer_list, customer) //if they exit the room
+
+
+/obj/effect/warped_rune/rainbowspace/attack_hand(mob/living/user)
+	. = ..()
+	for(var/mob/living/carbon/human/customer in rune_turf)
+		customer.forceMove(locate(room_reservation.bottom_left_coords[1] + 3, room_reservation.bottom_left_coords[2] + 6, room_reservation.bottom_left_coords[3]))
+		customer_check(customer, TRUE)
+
+
+///Will delete the room when the rune is destroyed if no customer is left in the room.
+/obj/effect/warped_rune/rainbowspace/Destroy()
+	if(!length(customer_list))
+		qdel(room_reservation)
+	return ..()
+
+
+///anyone on the exit rune when it is used will be teleported to the rune that was used to teleport to the warped room
+/obj/effect/warped_room_exit/attack_hand(mob/living/user)
+	. = ..()
+	for(var/mob/living/carbon/human/customer in get_turf(src))
+		customer.forceMove(exit_turf)
+		do_sparks(3, FALSE, get_turf(src))
+		enter_rune.customer_check(customer, FALSE)
+
+	if(!length(enter_rune.customer_list) && !locate(enter_rune) in exit_turf) //deletes the room if the rune doesn't exist anymore and all customers have left
+		qdel(enter_rune.room_reservation)
+		qdel(src)
+		return ..()
```

Comment:
deletion will end the proc call
Are you sure you want to call the parent on a deleted object?